File: 3-leetcode/03-hard/js/719-find-k-th-smallest-pair-distance.ts
const smallestDistancePair = (nums: number[], k: number): number => {
  const max = nums.reduce((acc, val) => max(acc, val), -1);

  const arr = Array.from({ length: max + 1 }, () => 0);

  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      const cur = Math.abs(nums[i] - nums[j]);

      arr[cur]++;
    }
  }

  for (let i = 0; i <= max; i++) {
    k -= arr[i];

    if (k <= 0) {
      return i;
    }
  }

  return -1;
};

File: 3-leetcode/03-hard/cpp/719-find-k-th-smallest-pair-distance.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
  int smallestDistancePair(vector<int> &nums, int k)
  {
    int arraySize = nums.size();

    int maxElement = *max_element(nums.begin(), nums.end());

    vector<int> distanceBucket(maxElement + 1, 0);

    for (int i = 0; i < arraySize; ++i)
    {
      for (int j = i + 1; j < arraySize; ++j)
      {
        int distance = abs(nums[i] - nums[j]);

        distanceBucket[distance]++;
      }
    }

    for (int dist = 0; dist <= maxElement; dist++)
    {
      k -= distanceBucket[dist];

      if (k <= 0)
      {
        return dist;
      }
    }
    return -1; // Return -1 if no distance found, should not reach here
  }
};